repair/row_level.cc

```diff
@@ -283,40 +298,55 @@ using is_local_reader = bool_class<class is_local_reader_tag>;
+                auto ms = mutation_source([&cf] (
+                    schema_ptr s,
+                    reader_permit permit,
+                    const dht::partition_range& pr,
+                    const query::partition_slice& ps,
+                    tracing::trace_state_ptr,
+                    streamed_mutation::forwarding,
+                    mutation_reader::forwarding fwd_mr) {
+                    return cf.make_streaming_reader(std::move(s), std::move(permit), pr, ps, fwd_mr);
+                });
+                flat_mutation_reader_v2 rd(nullptr);
+                std::tie(rd, _reader_handle) = make_manually_paused_evictable_reader_v2(
                     std::move(ms),
                     _schema,
                     _permit,
                     _range,
                     _schema->full_slice(),
                     {},
                     mutation_reader::forwarding::no);
-            return rd;
-        } else {
-            // We can't have two permits with count resource for 1 repair.
-            // So we release the one on _permit so the only one is the one the
-            // shard reader will obtain.
-            _permit.release_base_resources();
-            return make_multishard_streaming_reader(db, _schema, _permit, [this] {
-                auto shard_range = _sharder.next();
-                if (shard_range) {
-                    return std::optional<dht::partition_range>(dht::to_partition_range(*shard_range));
-                }
-                return std::optional<dht::partition_range>();
-            });
+                return rd;
+            }
+            case read_strategy::multishard_split: {
+                // We can't have two permits with count resource for 1 repair.
+                // So we release the one on _permit so the only one is the one the
+                // shard reader will obtain.
+                _permit.release_base_resources();
+                return make_multishard_streaming_reader(db, _schema, _permit, [this] {
+                    auto shard_range = _sharder.next();
+                    if (shard_range) {
+                        return std::optional<dht::partition_range>(dht::to_partition_range(*shard_range));
+                    }
+                    return std::optional<dht::partition_range>();
+                });
+            }
+            case read_strategy::multishard_filter: {
+                // We can't have two permits with count resource for 1 repair.
+                // So we release the one on _permit so the only one is the one the
+                // shard reader will obtain.
+                _permit.release_base_resources();
+                return make_filtering_reader(make_multishard_streaming_reader(db, _schema, _permit, _range),
+                    [&remote_sharder, remote_shard](const dht::decorated_key& k) {
+                        return remote_sharder.shard_of(k.token()) == remote_shard;
+                    });
+            }
         }
```

Comment:
This won't compile on gcc, which allows enums to have non-enumerated values when considering control flow. So it considers it legal of the switch to not match anything (it will warn like clang) and then error on a function with no return value.

The way to fix it is to add an on_internal_error() (which is marked [[noreturn]]).

Reply:
fixed